File: shared-datacenter-proxies/Nodejs/filesystem.js
const stdUtil = require('util');
const fs = require('fs');
const writeFilePromisified = stdUtil.promisify(fs.writeFile)
const appendFilePromisified = stdUtil.promisify(fs.appendFile)
const util = require('./util');
const settings = require('./settings');

const errorFilename = 'failed_requests.txt'

const writeErrorToStdout = (error) => console.log(`ERROR: ${error}`);

module.exports = {
  writeErrorToStdout: writeErrorToStdout,

  readLines: (path) => {
    try {
      const content = fs.readFileSync(path).toString('utf-8');
      return content.split("\n").filter((url) => url !== '')
    } catch (e) {
      util.printAndExit('Could not read file: ' + e.message)
    }
  },

  readProxyMap: (path) => {
    const proxyList = module.exports.readLines(path);

    const proxyMap = [];
    for (const index in proxyList) {
      const proxyUrl = proxyList[index];
      const regexResult = settings.ProxyRegex.exec(proxyUrl);
      if (!regexResult || regexResult.groups === undefined) {
        proxyMap[settings.DefaultProxyIndexName] = proxyUrl;
        continue;
      }

      const country = regexResult.groups.country.toUpperCase();
      proxyMap[country] = proxyUrl;
    }

    return proxyMap;
  },

  writeErrorToFile: async(error) => {
    return await appendFilePromisified(errorFilename, error + "\n")
  },

  writeSuccessToFile: async(position, content) => {
    const fileName = `result_${position}.html`;

    return await writeFilePromisified(fileName, content)
  }
}

File: shared-datacenter-proxies/GoLang/filesystem.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"io"
	"os"
	"strings"
)

const errorFilename = "failed_requests.txt"
const successFilename = "result_%d.html"

func readProxyMap(path string) (map[string]string, error) {
	proxyList, err := readLines(path)
	if err != nil {
		return nil, err
	}

	proxyMap := make(map[string]string)
	for _, proxyUrl := range proxyList {
		matchResult := matchProxyUrl(proxyUrl)

		if country, ok := matchResult["country"]; ok {
			proxyMap[strings.ToUpper(country)] = proxyUrl
		} else {
			proxyMap[DefaultProxyIndexName] = proxyUrl
		}
	}

	return proxyMap, nil
}

func readLines(path string) ([]string, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines, scanner.Err()
}

func writeErrorToStdout(format string, a ...interface{}) {
	fmt.Printf(format+"\n", a...)
}

func writeErrorToFile(format string, a ...interface{}) {
	message := fmt.Sprintf(format+"\n", a...)
	writeToFile(errorFilename, message)
}

func writeSuccessToFile(position int, responseStream io.Reader) {
	contents := streamToString(responseStream)
	formattedFilename := fmt.Sprintf(successFilename, position)
	writeToFile(formattedFilename, contents)
}

func streamToString(stream io.Reader) string {
	buf := new(bytes.Buffer)
	buf.ReadFrom(stream)
	return buf.String()
}

func writeToFile(filename, text string) {
	f, err := os.OpenFile(filename, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
	if err != nil {
		panic(err)
	}

	defer f.Close()

	if _, err = f.WriteString(text); err != nil {
		panic(err)
	}
}
